lib/notificationhub.go

```diff
@@ -0,0 +1,295 @@
+	if err != nil {
+		return nil, fmt.Errorf("notificationHub.Send: %s", err)
+	}
+	return b, nil
+}
+
+// SendDirect publishes notification to a specific device
+func (h *NotificationHub) SendDirect(ctx context.Context, n *Notification, deviceHandle string) ([]byte, error) {
+	b, err := h.sendDirect(ctx, n, deviceHandle)
+	if err != nil {
+		return nil, fmt.Errorf("notificationHub.SendDirect: %s", err)
+	}
+	return b, nil
+}
+
+// Schedule pusblishes a scheduled notification
+// Format tags according to https://docs.microsoft.com/en-us/azure/notification-hubs/notification-hubs-tags-segment-push-message
+// or nil if no tags should be used
+func (h *NotificationHub) Schedule(ctx context.Context, n *Notification, tags *string, deliverTime time.Time) ([]byte, error) {
+	b, err := h.send(ctx, n, tags, &deliverTime)
+	if err != nil {
+		return nil, fmt.Errorf("notificationHub.Schedule: %s", err)
+	}
+
+	return b, nil
+}
+
+// Registration reads one specific registration
+func (h *NotificationHub) Registration(ctx context.Context, deviceID string) (*RegistrationResult, []byte, error) {
+	var (
+		result = &RegistrationResult{}
+		regURL = h.generateAPIURL("registrations")
+	)
+	regURL.Path = path.Join(regURL.Path, deviceID)
+	rawResponse, err := h.exec(ctx, getMethod, regURL, Headers{}, nil)
+	if err != nil {
+		return nil, rawResponse, err
+	}
+	if err = xml.Unmarshal(rawResponse, &result); err != nil {
+		return nil, rawResponse, err
+	}
+	result.RegistrationContent.normalize()
+	return result, rawResponse, nil
+}
+
+// Registrations reads all registrations
+func (h *NotificationHub) Registrations(ctx context.Context) (*Registrations, []byte, error) {
+	rawResponse, err := h.exec(ctx, getMethod, h.generateAPIURL("registrations"), Headers{}, nil)
+	if err != nil {
+		return nil, rawResponse, err
+	}
+	result := &Registrations{}
+	if err = xml.Unmarshal(rawResponse, &result); err != nil {
+		return nil, rawResponse, err
+	}
+	result.normalize()
+	return result, rawResponse, nil
+}
+
+// Register sends registration to the azure hub
+func (h *NotificationHub) Register(ctx context.Context, r Registration) (RegistrationResult, []byte, error) {
+	var (
+		regRes  = RegistrationResult{}
+		regURL  = h.generateAPIURL("registrations")
+		method  = postMethod
+		payload = ""
+		headers = map[string]string{
+			"Content-Type": "application/atom+xml;type=entry;charset=utf-8",
+		}
+	)
+
+	switch r.NotificationFormat {
+	case AppleFormat:
+		payload = strings.Replace(AppleRegTemplate, "{{DeviceID}}", r.DeviceID, 1)
+	case GcmFormat:
+		payload = strings.Replace(GcmRegTemplate, "{{DeviceID}}", r.DeviceID, 1)
+	default:
+		return regRes, nil, errors.New("Notification format not implemented")
+	}
+	payload = strings.Replace(payload, "{{Tags}}", r.Tags, 1)
+
+	if r.RegistrationID != "" {
+		method = putMethod
+		regURL.Path = path.Join(regURL.Path, r.RegistrationID)
+	}
+
+	res, err := h.exec(ctx, method, regURL, headers, bytes.NewBufferString(payload))
```

Comment:
Mixing abbreviated result and error, I think it's better to just pick one. For example res or result

Reply:
👍